### PR TITLE
MAINT: Make constant and edge mode padding faster.

### PR DESCRIPTION
**Pull request by hmaarrfk**
Kept only as reference, while https://github.com/numpy/numpy/pull/11358 is being developed.

This is a start at addressing #11126

Padding was causing the array to be copied `2*arr.ndim` times which can be huge for large arrays.

I did not comment my helper functions. 
They are 1-2 line long.
If numpy requires those to be commented then please let me know.

Modes finished:
  * [x] constant
  * [x] edge
  * [x] linear_ramp
  * [x] maximum
  * [x] mean
  * [x] median
  * [x] minimum
  * [ ] reflect
  * [ ] symmetric
  * [ ] wrap

I'm glad to pass on the torch.

Before
```
================== ========== ========== ============= ========== ========== ========== ========== ========== =========== ==========
--                                                                        mode                                                      
------------------ -----------------------------------------------------------------------------------------------------------------
    shape         constant     edge     linear_ramp   maximum      mean      median    minimum    reflect    symmetric     wrap   
================== ========== ========== ============= ========== ========== ========== ========== ========== =========== ==========
   (64, 64)       625±0μs    747±0μs      1.32±0ms    1.03±0ms   1.22±0ms   2.51±0ms   1.41±0ms   506±0μs     617±0μs    618±0μs  
  (256, 256)      2.57±0ms   3.02±0ms     4.44±0ms    4.56±0ms   4.50±0ms   9.95±0ms   4.61±0ms   2.69±0ms    2.71±0ms   2.59±0ms 
  (512, 512)      8.38±0ms   6.59±0ms     7.64±0ms    9.02±0ms   7.15±0ms   19.3±0ms   11.4±0ms   8.07±0ms    6.77±0ms   8.64±0ms 
 (10, 10, 10)     1.31±0ms   1.02±0ms     2.56±0ms    883±0μs    2.09±0ms   3.74±0ms   1.81±0ms   1.28±0ms    892±0μs    853±0μs  
 (10, 64, 64)     5.88±0ms   8.69±0ms     12.2±0ms    9.89±0ms   6.38±0ms   13.3±0ms   10.8±0ms   5.91±0ms    5.16±0ms   4.39±0ms 
(50, 1024, 1024)   1.88±0s    1.85±0s      2.20±0s     2.23±0s    2.15±0s    5.63±0s    2.24±0s    1.11±0s     1.10±0s    1.11±0s  
================== ========== ========== ============= ========== ========== ========== ========== ========== =========== ==========
```

After (only constant and edge were implemented)
```


================== ========== ========== ============= ========== ========== ========== ========== ========== =========== ==========
--                                                                        mode                                                      
------------------ -----------------------------------------------------------------------------------------------------------------
    shape         constant     edge     linear_ramp   maximum      mean      median    minimum    reflect    symmetric     wrap   
================== ========== ========== ============= ========== ========== ========== ========== ========== =========== ==========
   (64, 64)       840±0μs    825±0μs      916±0μs     962±0μs    858±0μs    2.10±0ms   667±0μs    595±0μs     655±0μs    634±0μs  
  (256, 256)      1.94±0ms   1.91±0ms     4.45±0ms    4.57±0ms   4.41±0ms   5.65±0ms   3.16±0ms   2.46±0ms    1.11±0ms   2.57±0ms 
  (512, 512)      3.00±0ms   4.77±0ms     10.4±0ms    10.4±0ms   9.34±0ms   21.4±0ms   10.6±0ms   6.99±0ms    6.77±0ms   4.52±0ms 
 (10, 10, 10)     1.10±0ms   1.14±0ms     2.57±0ms    1.83±0ms   1.49±0ms   3.84±0ms   1.83±0ms   1.30±0ms    938±0μs    605±0μs  
 (10, 64, 64)     2.66±0ms   3.33±0ms     10.7±0ms    8.38±0ms   7.58±0ms   14.5±0ms   8.83±0ms   5.67±0ms    4.82±0ms   5.01±0ms 
(50, 1024, 1024)   411±0ms    419±0ms      2.20±0s     2.23±0s    2.15±0s    5.69±0s    2.25±0s    1.12±0s     1.11±0s    1.08±0s  
================== ========== ========== ============= ========== ========== ========== ========== ========== =========== ==========

```

cc @jakirkham

### CODE REVIEW

File: benchmarks/benchmarks/bench_pad.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+from __future__ import absolute_import, division, print_function
+
+from .common import Benchmark
+
+import numpy as np
+
+
+class PadSuite(Benchmark):
```
Comment (eric-wieser) [inline benchmarks/benchmarks/bench_pad.py]: There's already a benchmark for this in `bench_lib.py`

**Comment by eric-wieser** [comment]
This seems to overlap heavily with the goals of #11358

**Comment by hmaarrfk** [comment]
Man. Totally didn’t see that PR. Ill have to compare as to how it fairs with the performance of this one.

**Comment by eric-wieser** [comment]
I think I'd like to see a version of this built on top of #11955 - the one-copy approach is neat, but I think everything would become less convoluted if we could separate calculating the padded values from inserting them

**Comment by hmaarrfk** [comment]
Hey Eric. 

Thanks for having a look ar my PR. It is a little late for me but ill try to take a look at your PR tomorrow.

Mark

File: benchmarks/benchmarks/bench_pad.py (hunk 1 above)
Comment (hmaarrfk) [reply benchmarks/benchmarks/bench_pad.py]: Done! I made a small modification to it and added some common use cases.

**Comment by hmaarrfk** [comment]
@lagru I think I'm mostly done playing with this puzzle. I'm really not interested in re-implementing the  "symmetric", "reflect" and "warp" modes myself, especially since you seem to have implemented them nicely.

Let me know when you finish committing your changes, I think there is definitely a nice intersection of changes that we can whip up.

My benchmark actually shows that I slow down small array performance by a factor of 2.

<details>
<summary>Benchmark results</summary>

```
       before           after         ratio
     [58da5687]       [106a9a1d]
     <master>         <pad_single_copy>
+      41.0±0.9μs        82.6±10μs     2.01  bench_lib.Pad.time_pad((10, 100), 1, 'edge')
+      31.5±0.2μs       62.5±0.1μs     1.98  bench_lib.Pad.time_pad((1000,), (0, 5), 'minimum')
+      24.5±0.6μs         48.2±1μs     1.96  bench_lib.Pad.time_pad((1000,), (0, 5), 'edge')
+      33.9±0.9μs         65.2±2μs     1.92  bench_lib.Pad.time_pad((1000,), (0, 5), 'maximum')
+      38.8±0.4μs       74.3±0.7μs     1.92  bench_lib.Pad.time_pad((1000,), 1, 'maximum')
+      38.7±0.2μs         73.9±2μs     1.91  bench_lib.Pad.time_pad((1000,), 3, 'minimum')
+      39.7±0.9μs         75.6±2μs     1.91  bench_lib.Pad.time_pad((1000,), 3, 'maximum')
+     26.5±0.08μs      50.4±0.07μs     1.90  bench_lib.Pad.time_pad((1000,), 1, 'edge')
+      69.8±0.4μs        132±0.9μs     1.90  bench_lib.Pad.time_pad((1000,), 3, 'linear_ramp')
+      39.0±0.3μs       73.9±0.4μs     1.90  bench_lib.Pad.time_pad((1000,), 1, 'minimum')
+      47.1±0.2μs       88.2±0.7μs     1.87  bench_lib.Pad.time_pad((10, 100), (0, 5), 'minimum')
+      53.2±0.8μs         98.7±2μs     1.86  bench_lib.Pad.time_pad((1000,), (0, 5), 'linear_ramp')
+       124±0.7μs          230±2μs     1.85  bench_lib.Pad.time_pad((10, 100), 3, 'linear_ramp')
+        27.6±1μs       50.6±0.3μs     1.84  bench_lib.Pad.time_pad((1000,), 3, 'edge')
+      84.1±0.4μs        154±0.6μs     1.83  bench_lib.Pad.time_pad((10, 100), (0, 5), 'linear_ramp')
+         134±1μs          244±7μs     1.83  bench_lib.Pad.time_pad((64, 64), 3, 'linear_ramp')
+      47.6±0.2μs       86.8±0.3μs     1.82  bench_lib.Pad.time_pad((10, 100), (0, 5), 'maximum')
+      64.1±0.1μs        117±0.3μs     1.82  bench_lib.Pad.time_pad((10, 100), 1, 'minimum')
+        64.7±1μs          118±1μs     1.82  bench_lib.Pad.time_pad((10, 100), 1, 'maximum')
+        68.9±2μs        125±0.6μs     1.81  bench_lib.Pad.time_pad((10, 100), 3, 'minimum')
+       203±0.3μs         365±10μs     1.80  bench_lib.Pad.time_pad((10, 10, 10), 3, 'linear_ramp')
+         127±2μs          224±2μs     1.77  bench_lib.Pad.time_pad((10, 10, 10), (0, 5), 'linear_ramp')
+      55.9±0.5μs         98.9±3μs     1.77  bench_lib.Pad.time_pad((10, 10, 10), 1, 'edge')
+         103±2μs          182±6μs     1.76  bench_lib.Pad.time_pad((10, 10, 10), 1, 'minimum')
+      76.2±0.4μs          134±1μs     1.76  bench_lib.Pad.time_pad((10, 10, 10), (0, 5), 'minimum')
+        92.8±3μs          162±3μs     1.75  bench_lib.Pad.time_pad((64, 64), (0, 5), 'linear_ramp')
+      41.1±0.5μs       71.6±0.3μs     1.74  bench_lib.Pad.time_pad((10, 100), 3, 'edge')
+      43.7±0.3μs         75.4±3μs     1.73  bench_lib.Pad.time_pad((1000,), (0, 5), 'mean')
+      35.2±0.2μs       60.4±0.3μs     1.72  bench_lib.Pad.time_pad((10, 100), (0, 5), 'edge')
+         108±4μs          184±6μs     1.69  bench_lib.Pad.time_pad((10, 10, 10), 1, 'maximum')
+      83.5±0.7μs          141±8μs     1.69  bench_lib.Pad.time_pad((64, 64), 1, 'maximum')
+        48.3±1μs         81.3±2μs     1.69  bench_lib.Pad.time_pad((10, 10, 10), (0, 5), 'edge')
+        69.5±1μs        117±0.3μs     1.69  bench_lib.Pad.time_pad((10, 100), 3, 'maximum')
+      59.1±0.3μs       96.9±0.4μs     1.64  bench_lib.Pad.time_pad((64, 64), (0, 5), 'minimum')
+      59.7±0.5μs       97.3±0.5μs     1.63  bench_lib.Pad.time_pad((64, 64), (0, 5), 'maximum')
+         123±5μs          198±7μs     1.61  bench_lib.Pad.time_pad((10, 10, 10), 3, 'minimum')
+      79.6±0.9μs          128±3μs     1.61  bench_lib.Pad.time_pad((10, 10, 10), (0, 5), 'maximum')
+        85.2±2μs          136±5μs     1.60  bench_lib.Pad.time_pad((64, 64), 1, 'minimum')
+      33.8±0.1μs       53.9±0.2μs     1.59  bench_lib.Pad.time_pad((1000,), (0, 5), 'constant')
+      47.2±0.1μs         75.0±2μs     1.59  bench_lib.Pad.time_pad((64, 64), 1, 'edge')
+        40.5±2μs       64.3±0.6μs     1.59  bench_lib.Pad.time_pad((64, 64), (0, 5), 'edge')
+         124±2μs          193±2μs     1.55  bench_lib.Pad.time_pad((10, 10, 10), 3, 'maximum')
+        50.1±2μs         77.6±2μs     1.55  bench_lib.Pad.time_pad((64, 64), 3, 'edge')
+        66.8±2μs          102±5μs     1.53  bench_lib.Pad.time_pad((10, 10, 10), 3, 'edge')
+      66.7±0.5μs        102±0.5μs     1.53  bench_lib.Pad.time_pad((10, 100), (0, 5), 'mean')
+      86.6±0.3μs        132±0.8μs     1.53  bench_lib.Pad.time_pad((64, 64), 3, 'maximum')
+        88.1±3μs        132±0.8μs     1.50  bench_lib.Pad.time_pad((64, 64), 3, 'minimum')
+      60.3±0.7μs         89.2±1μs     1.48  bench_lib.Pad.time_pad((1000,), 3, 'mean')
+     37.0±0.06μs       54.3±0.2μs     1.47  bench_lib.Pad.time_pad((1000,), 1, 'constant')
+        63.8±3μs         93.3±1μs     1.46  bench_lib.Pad.time_pad((1000,), 1, 'mean')
+        38.0±2μs         54.9±1μs     1.45  bench_lib.Pad.time_pad((1000,), 3, 'constant')
+       101±0.3μs        144±0.7μs     1.43  bench_lib.Pad.time_pad((10, 100), 1, 'mean')
+       102±0.8μs          144±4μs     1.41  bench_lib.Pad.time_pad((10, 100), 3, 'mean')
+      49.2±0.3μs         68.6±2μs     1.39  bench_lib.Pad.time_pad((64, 64), (0, 5), 'constant')
+        78.1±2μs          109±6μs     1.39  bench_lib.Pad.time_pad((64, 64), (0, 5), 'mean')
+         107±2μs        149±0.8μs     1.39  bench_lib.Pad.time_pad((10, 10, 10), (0, 5), 'mean')
+        85.4±2μs        117±0.7μs     1.37  bench_lib.Pad.time_pad((1000,), (0, 5), 'median')
+        46.7±2μs       63.8±0.2μs     1.37  bench_lib.Pad.time_pad((10, 100), (0, 5), 'constant')
+         158±7μs          215±2μs     1.36  bench_lib.Pad.time_pad((10, 10, 10), 1, 'mean')
+         249±7μs         335±30μs     1.35  bench_lib.Pad.time_pad((10, 10, 10), (0, 5), 'median')
+        58.3±2μs         77.3±2μs     1.33  bench_lib.Pad.time_pad((10, 10, 10), (0, 5), 'constant')
+         180±5μs         235±20μs     1.31  bench_lib.Pad.time_pad((64, 64), (0, 5), 'median')
+         119±4μs          153±4μs     1.29  bench_lib.Pad.time_pad((64, 64), 3, 'mean')
+         176±3μs          227±1μs     1.29  bench_lib.Pad.time_pad((10, 10, 10), 3, 'mean')
+      52.9±0.4μs       68.3±0.5μs     1.29  bench_lib.Pad.time_pad((10, 100), 3, 'constant')
+         151±3μs          194±5μs     1.28  bench_lib.Pad.time_pad((10, 100), (0, 5), 'median')
+      40.3±0.4μs       50.7±0.2μs     1.26  bench_lib.Pad.time_pad((64, 64), (0, 5), 'reflect')
+        60.1±1μs       75.2±0.3μs     1.25  bench_lib.Pad.time_pad((64, 64), 1, 'constant')
+        417±10μs         506±10μs     1.21  bench_lib.Pad.time_pad((10, 10, 10), 1, 'median')
+         138±4μs        165±0.5μs     1.19  bench_lib.Pad.time_pad((1000,), 3, 'median')
+         272±8μs          325±4μs     1.19  bench_lib.Pad.time_pad((10, 100), 3, 'median')
+         139±1μs          164±2μs     1.18  bench_lib.Pad.time_pad((1000,), 1, 'median')
+        70.8±1μs         83.1±1μs     1.17  bench_lib.Pad.time_pad((10, 10, 10), 1, 'constant')
+        75.9±1μs         88.6±3μs     1.17  bench_lib.Pad.time_pad((10, 10, 10), 3, 'constant')
+         445±2μs          514±3μs     1.16  bench_lib.Pad.time_pad((10, 10, 10), 3, 'median')
+         272±1μs         312±10μs     1.15  bench_lib.Pad.time_pad((10, 100), 1, 'median')
+       299±0.5μs          342±2μs     1.14  bench_lib.Pad.time_pad((64, 64), 1, 'median')
+         305±1μs        348±0.7μs     1.14  bench_lib.Pad.time_pad((64, 64), 3, 'median')
+        63.4±3μs       71.2±0.3μs     1.12  bench_lib.Pad.time_pad((64, 64), 3, 'constant')
-         154±1ms          133±1ms     0.86  bench_lib.Pad.time_pad((10, 512, 512), 1, 'median')
-       110±0.6ms         90.4±2ms     0.82  bench_lib.Pad.time_pad((10, 512, 512), (0, 5), 'median')
-         258±2ms          192±5ms     0.74  bench_lib.Pad.time_pad((10, 512, 512), 3, 'median')
-      6.11±0.2ms       4.40±0.2ms     0.72  bench_lib.Pad.time_pad((512, 512), (0, 5), 'median')
-      3.23±0.2ms       1.82±0.5ms     0.56  bench_lib.Pad.time_pad((512, 512), (0, 5), 'minimum')
-      3.23±0.2ms       1.81±0.5ms     0.56  bench_lib.Pad.time_pad((512, 512), (0, 5), 'maximum')
-      3.16±0.2ms       1.75±0.5ms     0.55  bench_lib.Pad.time_pad((512, 512), (0, 5), 'mean')
-      2.92±0.2ms       1.56±0.5ms     0.54  bench_lib.Pad.time_pad((512, 512), (0, 5), 'linear_ramp')
-      4.46±0.2ms       2.26±0.5ms     0.51  bench_lib.Pad.time_pad((512, 512), 1, 'maximum')
-      2.78±0.2ms       1.41±0.5ms     0.50  bench_lib.Pad.time_pad((512, 512), (0, 5), 'edge')
-      2.77±0.2ms       1.39±0.5ms     0.50  bench_lib.Pad.time_pad((512, 512), (0, 5), 'constant')
-      4.21±0.2ms       2.11±0.5ms     0.50  bench_lib.Pad.time_pad((512, 512), 3, 'mean')
-      4.46±0.2ms       2.22±0.5ms     0.50  bench_lib.Pad.time_pad((512, 512), 3, 'minimum')
-      4.43±0.2ms       2.20±0.5ms     0.50  bench_lib.Pad.time_pad((512, 512), 1, 'minimum')
-      4.47±0.2ms       2.22±0.5ms     0.50  bench_lib.Pad.time_pad((512, 512), 3, 'maximum')
-      4.18±0.2ms       2.03±0.5ms     0.49  bench_lib.Pad.time_pad((512, 512), 1, 'mean')
-      57.8±0.4ms       27.4±0.1ms     0.47  bench_lib.Pad.time_pad((10, 512, 512), 1, 'minimum')
-      60.2±0.9ms       28.5±0.8ms     0.47  bench_lib.Pad.time_pad((10, 512, 512), 1, 'maximum')
-      3.81±0.2ms       1.70±0.5ms     0.45  bench_lib.Pad.time_pad((512, 512), 3, 'linear_ramp')
-        44.1±2ms       19.5±0.1ms     0.44  bench_lib.Pad.time_pad((10, 512, 512), (0, 5), 'maximum')
-      45.3±0.8ms       19.5±0.4ms     0.43  bench_lib.Pad.time_pad((10, 512, 512), (0, 5), 'minimum')
-      51.0±0.7ms       21.7±0.2ms     0.43  bench_lib.Pad.time_pad((10, 512, 512), 1, 'mean')
-      3.50±0.2ms       1.44±0.5ms     0.41  bench_lib.Pad.time_pad((512, 512), 1, 'edge')
-        39.6±2ms       16.1±0.3ms     0.41  bench_lib.Pad.time_pad((10, 512, 512), (0, 5), 'mean')
-      3.67±0.2ms       1.49±0.5ms     0.41  bench_lib.Pad.time_pad((512, 512), 1, 'linear_ramp')
-      3.52±0.2ms       1.42±0.5ms     0.40  bench_lib.Pad.time_pad((512, 512), 3, 'edge')
-      3.51±0.2ms       1.41±0.5ms     0.40  bench_lib.Pad.time_pad((512, 512), 3, 'constant')
-      3.66±0.2ms       1.41±0.5ms     0.38  bench_lib.Pad.time_pad((512, 512), 1, 'constant')
-       110±0.2ms         40.7±1ms     0.37  bench_lib.Pad.time_pad((10, 512, 512), 3, 'minimum')
-         116±7ms       39.1±0.1ms     0.34  bench_lib.Pad.time_pad((10, 512, 512), 3, 'maximum')
-         102±1ms       32.5±0.5ms     0.32  bench_lib.Pad.time_pad((10, 512, 512), 3, 'mean')
-      50.0±0.9ms      10.4±0.07ms     0.21  bench_lib.Pad.time_pad((10, 512, 512), (0, 5), 'linear_ramp')
-         102±2ms       18.9±0.4ms     0.19  bench_lib.Pad.time_pad((10, 512, 512), 3, 'linear_ramp')
-        44.5±2ms      7.58±0.02ms     0.17  bench_lib.Pad.time_pad((10, 512, 512), (0, 5), 'edge')
-        89.2±3ms       15.1±0.3ms     0.17  bench_lib.Pad.time_pad((10, 512, 512), 3, 'edge')
-      42.7±0.5ms      7.05±0.04ms     0.17  bench_lib.Pad.time_pad((10, 512, 512), (0, 5), 'constant')
-        87.5±2ms      13.5±0.06ms     0.15  bench_lib.Pad.time_pad((10, 512, 512), 3, 'constant')
-      42.8±0.9ms      6.58±0.08ms     0.15  bench_lib.Pad.time_pad((10, 512, 512), 1, 'edge')
-        42.6±1ms      6.31±0.09ms     0.15  bench_lib.Pad.time_pad((10, 512, 512), 1, 'constant')
-      58.5±0.4ms      7.06±0.02ms     0.12  bench_lib.Pad.time_pad((10, 512, 512), 1, 'linear_ramp')
```
</details>

**Comment by hmaarrfk** [comment]
Closed in favour of #11358

**Pull request closed without merging**